[1/N] Move NaN check onto NCCL stream

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134707
* #134701
* #134357
* #134345
* __->__ #134300

So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

Also moved declaration of `checkForNan` from Utils.hpp to NCCLUtils.hpp, and renamed Utils.cu to NCCLUtils.cu.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D61957573](https://our.internmc.facebook.com/intern/diff/D61957573)

kwen2501 committed:
[1/N] Move NaN check onto NCCL stream

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2667,6 +2664,10 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
 
   at::cuda::OptionalCUDAGuard gpuGuard;
```
did you determine what the 'gpuGuard' does and why it matters (or is it not important)

wconstab replied:
(I know CudaGuard(device_id) can be used to set the device.  What does it mean if you use OptionalCUDAGuard and do not set the device?)

kwen2501 replied:
Likely just creating a CUDA context if there isn't one already.
If no device id is given, default device would be used.
See: 
https://pytorch.org/cppdocs/api/structc10_1_1cuda_1_1_optional_c_u_d_a_guard.html
https://pytorch.org/cppdocs/api/structc10_1_1cuda_1_1_c_u_d_a_guard.html#structc10_1_1cuda_1_1_c_u_d_a_guard

shuqiangzhang replied:
Yes, we might need to set the device. Similar 'GPU0' bugs related to NCCL abort (https://github.com/pytorch/pytorch/pull/127363/), we just set the context at the PTD level as a mitigation

wconstab commented:
I want the PR desc updated to clarify 

1) moving to the nccl stream is a preference we think is better not a requirement
2) ensuring the nan check is run on the proper gpu (via deviceguard) solves the bug

(im not sure if what i said in 1, 2 are even true but i want your PR desc to educate me about those details)

wconstab commented:
can you also add tests? i think you can put a couple tests on each PR that cover the changes that PR fixed.

kwen2501 commented:
> 1. moving to the nccl stream is a preference we think is better not a requirement

Agree

> 2. ensuring the nan check is run on the proper gpu (via deviceguard) solves the bug

Indeed, this PR didn't fix anything related to the device to run the NaN kernel.
Maybe there will be a third PR.

Also, I wonder if
```
at::cuda::OptionalCUDAGuard gpuGuard;
```
is the cause of additional context on GPU 0 by **all** processes when one type nvidia-smi.

kwen2501 committed:
Update on "[1/N] Move NaN check onto NCCL stream"


So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[1/N] Move NaN check onto NCCL stream"


So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

Also moved declaration of `checkForNan` from Utils.hpp to NCCLUtils.hpp, and renamed Utils.cu to NCCLUtils.cu.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
> can you also add tests? i think you can put a couple tests on each PR that cover the changes that PR fixed.

Relying on original test as test for this PR as this PR is a "cosmetic" change -- original test should continue to work.

shuqiangzhang reviewed (approved):
seeing the test passing in the 3rd PR

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot --help

kwen2501 commented:
@pytorchbot revert -m "This is breaking builds of MTIA"

kwen2501 commented:
@pytorchbot revert -m "This is breaking builds of MTIA" -c weird

kwen2501 committed:
Update on "[1/N] Move NaN check onto NCCL stream"


So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

Also moved declaration of `checkForNan` from Utils.hpp to NCCLUtils.hpp, and renamed Utils.cu to NCCLUtils.cu.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[1/N] Move NaN check onto NCCL stream"


So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

Also moved declaration of `checkForNan` from Utils.hpp to NCCLUtils.hpp, and renamed Utils.cu to NCCLUtils.cu.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[1/N] Move NaN check onto NCCL stream"


So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

Also moved declaration of `checkForNan` from Utils.hpp to NCCLUtils.hpp, and renamed Utils.cu to NCCLUtils.cu.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[1/N] Move NaN check onto NCCL stream"


So that the tensor's lifetime management is the same as the management built for the NCCL, pre and post kernels.
Also so that on visualizers, they show up in the NCCL stream line. Otherwise if they show up in the compute line, user may get confused (my code does not have these kernels).

The check is thus moved after the point where we depend NCCL stream from the last compute kernel.

Also moved declaration of `checkForNan` from Utils.hpp to NCCLUtils.hpp, and renamed Utils.cu to NCCLUtils.cu.

cc XilunWu H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61957573).

kwen2501 commented:
@pytorchbot merge
